designs/Transitive-Dependencies.md

```diff
@@ -0,0 +1,150 @@
+There is not a great solution that exists to understand the nature of top-level packages and their transitive dependencies.
+The solution explorer in Visual Studio does the best job at allowing a user to dive into each NuGet package & see all transitive dependencies from the top-level dependency.
+
+![](../resources/TransitiveDependencies/SolutionView.png)
+
+The dotnet CLI does not provide the insight into “why” a transitive dependency is listed, although it does list everything that has been resolved.
+
+![](../resources/TransitiveDependencies/DotNetCLI.png)
+
+The Visual Studio Package Manager UI does not support the concept of showing transitive packages in any fashion.
+
+![](../resources/TransitiveDependencies/VSPMUI.png)
+
+## Success: How do we know if we have solved this problem?
+
+By measuring the number of detractors regarding transitive dependencies & promoters of this new experience in future NuGet client surveys.
+
+We will add telemetry to categorize when a top-level package is installed, and whether it was done in the context of promoting a transitive package to a top-level one.
+
+## Audience: Who are we building this for?
+
+We are building this for .NET users, both old and new, to be successful with understanding the types of NuGet dependencies in their project(s) and solution(s).
+
+## What: What does this look like in the product?
+
+![](../resources/TransitiveDependencies/TransitiveVSPMUI.png)
+
+## Context and Scope
+
+-	https://github.com/NuGet/Home/issues/5887 
+-	https://developercommunity.visualstudio.com/content/idea/538119/add-dependency-tree-view-for-packagereferences.html 
+
+## Minimal Requirements
+
+-	Visual list of transitive dependencies.
+-	Ability to see a transitive dependency package details.
+-	Ability to determine where a transitive dependency originated.
+-	Ability to promote a transitive dependency to a top-level dependency.
+
+## Scenarios and User Experience
+
+### CLI
+
+We should consider a minor tweak to the existing experience of “dotnet list package --include-transitive” to provide the user with a sense of where the package came from. This could be a new column next to “Resolved” which says “Transitively Referenced” or “Referenced” and has a list of the top-level packages that requested the dependency.
+
+![](../resources/TransitiveDependencies/TransitiveDotNetCLI.png)
+
+### Visual Studio
+
+Within Visual Studio, there will be a new panel within the Visual Studio Package Manager UI named “Transitive Packages” in which all transitive packages will be displayed to the user. For the sake of not confusing the user, there will additionally be a header titled “Top-level Packages” where the current experience is today for top-level packages.
+
+**By default, a user’s transitive packages will be collapsed** as a project can have hundreds, if not thousands of transitive packages & would potentially impact performance displaying these many records on each load.
```

Comment:
Are we going to audit transitive packages automatically and display that in the updates/installed labels? 

That's what we're doing for the top level packages. 

Note that I'm not recommending this, merely just clarifying :)

Reply:
Mind clarifying a bit more on "audit transitive packages automatically"? Just so I understand further!

Reply:
Currently with the top level only approach, the updates tab will have a label telling you the amount of packages that have updates (this is pretty bad perf right now). 

When deprecation was added, a warning icon was equivalent to the number in the Updates tab on the Installed tab.
This means that when we look-up the metadata for the packages installed to determine if they have any newer versions, we also look at the metadata to determine whether they are deprecated. 

Are we going to do the equivalent for transitive dependencies?
Are we going to preemptively look for the metadata for the transitive dependencies to determine if they are deprecated/vulnerable? 

In particular I'm asking this because given the way the server protocol is written right now, this could lead to 100s of different network calls.

Reply:
What is the perf potential of 100s of network calls?

You would see the same behavior for any installed package whether top-level or transitive.

Installed tab would show the current version & the available version.

Update tab would show the current version & the available version.

We would only count top-level packages for the badge on the tabs given we don't want users to promote to top-level just for a new version.

Reply:
We usually throttle based on the CPUs. 
The impact would be delayed population of the information in the UI. 

The number of network calls NuGet makes to load the information needed for UI is a problem today for the solution tab where the package count is large. A rough guestimate would be that the project level PM UI now needs as much information as the solution level one. 

If we don't want to shown badges for top level, then these concerns go away :) 

Note that these perf problems should be somewhat solvable, it's just a resourcing concern that'd likely take more effort than just displaying the transitive dependencies :)

Reply:
It's also completely fine to just not show updates available for transitive either. It would be a nice parity thing to reuse the current experience to show if a resolved transitive has an update with the hover experience. If it's complex perf wise, we can cut it.

Reply:
I think that's perfectly reasonable. 

I basically want us to make progress on the basics in this area without having to worry about perf implications now. 

If we want to have full parity later, we'll judge it on it's own merit.

Reply:
The whole point of implementing transitive dependencies is for the metadata, right? Don't we want to see which transitive dependencies are vulnerable?

Reply:
My understanding at this time, is that we're chiefly implementing transitive dependencies in PMUI *specifically* for the metadata, right? Isn't the North Star here being able to see which transitive packages have vulnerabilities?

Reply:
(I do know that's not a part of this spec, but since it's come up in conversation, and we want a design that supports these future features, let's discuss it)

Reply:
I'm deferring to you both regarding what the complexities would be. The goal is to show transitive dependencies enough to unblock package vulnerability while keeping a consistent experience that we have today  🏃

Reply:
The `perf` concern I have is if we aggressively fetch the metadata. 

ie, do we fetch metadata for transitive dependencies if the transitive dependencies are collapsed?

Right now, if you go to the browse tab, we'll fetch the metadata for all packages on that view...just so we can show the label.

Reply:
I'm looking to take a step back and consider the intention driving this feature, and whether we're still achieving this objective by not fetching metadata for a collapsed list. If a customer is viewing the installed packages for their project and sees a collapsed transitive dependencies list, and we haven't yet parsed vulnerability or deprecation metadata for the dependency graph, the customer may be misled into thinking the project's only vulnerabilities (for example) are in the top-level packages, which brings us back to the product's trustworthiness. 

The way I see it, there are 3 paths forward:
- Leave transitive dependency metadata unfetched, transitive dependencies collapsed, therefore incomplete vulnerability/deprecation data for the customer (until the list is expanded). We could add a disclaimer on the collapsed list indicating that expanding the list is the only way to get a complete picture of project deprecation/vulnerability status, and that this may take some time to fetch. This will however give an inaccurate label on the tab if we wish to indicate vulnerability/deprecation status there.

- Fetch all the data for the graph up-front and take the perf hit.

- Rearchitect our services approach to provide a more performant (complete) picture up front.

Reply:
I'm going to merge this spec as per other approvals & we'll take up these questions at implementation or technical spec/design time.